### ISSUE TITLE
False Negative: LeftCurlyCheck

### ISSUE THREAD

**Issue by MUzairS15** [checkstyle/checkstyle#11410]
I have read check documentation: https://checkstyle.sourceforge.io/config_blocks.html#LeftCurly
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

default `option` value is `eol`.

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
```
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="LeftCurly"/>
    </module>

</module>
```
```
/var/tmp $ cat Test.java
public class Test {

    void method1 ()
    /* violation but not reported */ { }

     void method2 ()
     { } /* violation and reported */
}
```
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:7:6: '{' at column 6 should be on the previous line. [LeftCurly]
Audit done.
Checkstyle ends with 1 errors.
```

---
**Describe what you expect in detail.**
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:4:38: '{' at column 38 should be on the previous line. [LeftCurly]
[ERROR] /var/tmp/Test.java:7:6: '{' at column 6 should be on the previous line. [LeftCurly]
Audit done.
Checkstyle ends with 2 errors.

**Comment by nick-mancuso** [comment]
Fix for this should be to update check to use AST based approach only. @strkkk I feel like we have an issue opened for updating all checks like this, but I cannot find it. Do you have an idea of issue title?

**Comment by OttoKaaij** [comment]
@nick-mancuso I think you mean this? https://github.com/checkstyle/checkstyle/issues/11166

**Comment by rnveach** [comment]
The issue seems to be the code is looking directly at the source file and not going based on our ASTs in this scenario.

Since the default is `EOL`, it appears the code is looking at the source file for all whitespace before the curly to declare it must be on the previous line.
See: https://github.com/checkstyle/checkstyle/blob/a8ac66635e2c5a0eb70d32b4e88d0715addfb04c/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java#L471

I am not sure how the code came to be like this, but I assume it is because of our current limitation to know what is the last AST before the `{`. The code is not sure where the method actually ends and the brace begins and so we think if we find non-whitespace before the curly, that we are still on the same line as the method definition.
Something like:
````
public void method(
   String manyArguments) {
````

This is why method1 has no violation but method2 does.

> I think you mean this? https://github.com/checkstyle/checkstyle/issues/11166

This doesn't fall under this as the method used is `getLine` and not `getFileContents`.

**Comment by romani** [comment]
To catch this violation Check should be updated to be comments aware.
Such code is very artificial, nobody write code like this :) (multiline comment in the middle), so missed violation is ok. But very technically Yes it is better to do violation on this.

**Comment by MUzairS15** [comment]
Yes, we can check if there is a block comment, in addition, to checking for whitespace.

**Comment by MUzairS15** [comment]
```
src % cat Test.java
public class Test { void foo() { // violation expected
        }
    }

src % java -jar checkstyle-10.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.

**Comment by Kevin222004** [comment]
I am on it

**Comment by Kevin222004** [comment]
I am closing my pr.

Any contributor who is fixing this pr please keep this is mind https://github.com/checkstyle/checkstyle/pull/12548#issuecomment-1493355090 

suggested by @romani 
we confused by name https://checkstyle.org/property_types.html#LeftCurlyOption
EOL does not mean end of any line .
EOL means at the end of line where actual block is started.

Attention: We need to update doc of EOL to defined this nuance in meaning.

comments can be continue to be not used by Check.
We need update to use AST more in logic to stick { to method signature line.

```
void method1 ()
    /* violation but not reported */ { }
```

we should demand it to be:

```
void method1 () {
    /* violation but not reported */  }
```

**Comment by romani** [comment]
```
$ cat Test.java
public class Test {
    void method1 ()
    /* not reported */ { }
}

$ java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -t Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:1]
`--CLASS_DEF -> CLASS_DEF [1:1]
    |--MODIFIERS -> MODIFIERS [1:1]
    |   `--LITERAL_PUBLIC -> public [1:1]
    |--LITERAL_CLASS -> class [1:8]
    |--IDENT -> Test [1:14]
    `--OBJBLOCK -> OBJBLOCK [1:19]
        |--LCURLY -> { [1:19]
        |--METHOD_DEF -> METHOD_DEF [2:5]
        |   |--MODIFIERS -> MODIFIERS [2:5]
        |   |--TYPE -> TYPE [2:5]
        |   |   `--LITERAL_VOID -> void [2:5]
        |   |--IDENT -> method1 [2:10]
        |   |--LPAREN -> ( [2:18]
        |   |--PARAMETERS -> PARAMETERS [2:19]
        |   |--RPAREN -> ) [2:19]
        |   `--SLIST -> { [3:24]
        |       `--RCURLY -> } [3:26]
        `--RCURLY -> } [4:1]

```

we need to detect that this two nodes are not on same line:
```
        |   |--RPAREN -> ) [2:19]
        |   `--SLIST -> { [3:24]
```
